### ISSUE TITLE
Plans on adding a filter?

### ISSUE THREAD

**Issue by codeconomy** [adriaanvanrossum/saasforcovid.com#3]
Hi Adrian,

Awesome work! Have been sharing this far and wide!

Do you have plans on adding a filter to saasforcovid?

Reason: I was planning to create a local non-saas version of your project, but given the range of things I'd be listing I absolutely need a filter. I'd like to get a "no-drop down" type of filter in place. I've mocked up what I want to implement in the picture below (i.e. the filter would be always visible in full on the top of the screen, above the list of "posts", in  your case above the list of Saas products ).

So the question is: Do you intend implement a filter?

If not, would you be able to suggest how I'd implement it myself (not exactly a full-stack dev here:) )?

![filter_top_of_page](https://user-images.githubusercontent.com/41002226/82046663-36b9b100-96a9-11ea-88f5-2d90d818d3a5.png)

Once again, thanks for this good work! 👏🏻

PS: apologies for the horrendous choice of filter colors (didn't spend much time making a color palette for above sketch... ) and misalignment

**Comment by adriaandotcom** [comment]
Would love to have a filter and I'm happy to guide you in the right direction! I don't have more time for saasforcovid.com at the moment, so feel free to add a PR.

You want to update this function to filter entries:

https://github.com/adriaanvanrossum/saasforcovid.com/blob/9f227c607a383dae78b4c55430845cee3de24ca5/src/App.vue#L538-L541

Every entry has more or less these fields:

```js
{
  name: "Digital Ocean",
  approved": "Yes",
  category": "Hosting",
  url: "https://digitalocean.com",
  ...
}
```

To filter on the results you could do something like this:

```js
.filter(({ description, name, category }) => {
  if (!description || !name || !category) return false;
  if (this.category) return category === this.category;
  return true
}) 
```

Make sure to add `category` to the `data` object:

https://github.com/adriaanvanrossum/saasforcovid.com/blob/9f227c607a383dae78b4c55430845cee3de24ca5/src/App.vue#L514-L524

When this is done you can test it with setting a category manually.

After that works you can add fields above the page to show the filters. You can either do this by input field or by preset categories.

Let me know if you need any help.

**Comment by codeconomy** [comment]
Thanks a ton @adriaanvanrossum for the guidance! Really appreciated!

Following up: 

Three questions as I explore: 

1) On the data object: 
https://github.com/adriaanvanrossum/saasforcovid.com/blob/9f227c607a383dae78b4c55430845cee3de24ca5/src/App.vue#L514-L524

**Apologies, but what do you mean with adding category to the data object? In which way should I do that?:**

```
data() { 
     return { 
       json: services, 
       error: null, 
       services: [], 
       loading: true, 
       buildDate: BUILD_DATE ? format(BUILD_DATE) : null, 
       commitHash: COMMIT_HASH, 
       show: ["maybe", "no"], 
     }; 
   }, 

```
<del> 2) Tried to test early filtering by making a copy of your google sheet and before anything else checking if changes in my sheet would render on the UI. I've then edited `update-cache.js` changing it to my Google sheet URL in here: </del>
https://github.com/adriaanvanrossum/saasforcovid.com/blob/d287e9fcee340fbaa30fdaf22b5e9df9d032ad0e/lib/update-cache.js#L5

<del> and then setting up the two github actions workflow files you have. 
https://github.com/adriaanvanrossum/saasforcovid.com/blob/master/.github/workflows/deploy.yml
and https://github.com/adriaanvanrossum/saasforcovid.com/blob/master/.github/workflows/update-cache.yml </del>

<del> Despite that, if I do changes to the Google sheet, `cache.js` does not update itself, so the UI does not change. **Any pointers of what I may be doing wrong?** </del>

<del> (Before anything, I've also double-checked that the downloaded CSV from Google sheets is being modified as changes are made to the google sheet, so the csv is ok, but the `.get(SHEET)` does not seem to be working.) </del>

👆 The above is fixed. For others in future --> double check the link to your google sheet. 
Also, to update your `cache.js` on your local environment run `npm run cache`

3) <del>I cannot seem to be able to deploy to Github Pages. Nothing renders. **Any** **pointers**?
PS: I have no issues deploying to Netlify (but then I am not able to use github Actions and make updates from Google sheets)</del>

👆 The above is fixed. For others in future --> I've given up trying to make this work on Github pages' "github.io" URLs (the JS and CSS files were not being found given relative Github pages relative paths and consequently the site would not render (blank). Go setup a custom domain in your repo's setting and this deployment will work.

Hope those were clear, otherwise, please let me know and I'll revert asap.

**So @adriaanvanrossum should you have any pointer to share on #1 above do let me know 🤗**

Thanks a ton!

**Comment by adriaandotcom** [comment]
1. How about this?

```js
data() { 
  return { 
    json: services, 
    error: null, 
    services: [], 
    loading: true, 
    buildDate: BUILD_DATE ? format(BUILD_DATE) : null, 
    commitHash: COMMIT_HASH, 
    show: ["maybe", "no"], 
    category: null,
  }; 
}, 
```

3. What is going wrong with deploying to GitHub Pages?

**Comment by codeconomy** [comment]
Thanks for reverting @adriaanvanrossum 

Following up:

> 1. What is going wrong with deploying to GitHub Pages?

Github pages places relative paths on `index.html` in the `gh-pages` branch. So all css files and js files are not found if one is deploying the project using Github page's "github.io" free URLs. I've been able to fix this by deleting the first "/" from the path of all css and js files linked to `index.html` in the `gh-pages` branch. This however isn't permanent, given any new deployment (following a github action) again makes the index.html add a "/" to the beginning of the css and js links and the app again won't render.

In other words, the js file isn't here: https://codeconomy.github.io/js/app.12c89b73.js (as per `index.html`), it is here: https://codeconomy.github.io/codeconomy.github.io-list/js/app.12c89b73.js

I'm unsure of where this could be fixed in your project. <ins>Anyways, this is a non-issue should one use a custom domain, all links to the css files and js files work perfectly that way and the app renders.</ins>

This is the line in your project:

https://github.com/adriaanvanrossum/saasforcovid.com/blob/f837c0a1ddce24429410e68c682a0b6cba3e3ac2/index.html#L12

**Comment by codeconomy** [comment]
Back to the filters:



> 1. How about this?
> 
> ```js
> data() { 
>   return { 
>     json: services, 
>     error: null, 
>     services: [], 
>     loading: true, 
>     buildDate: BUILD_DATE ? format(BUILD_DATE) : null, 
>     commitHash: COMMIT_HASH, 
>     show: ["maybe", "no"], 
>     category: null,
>   }; 
> }, 
> ```

This plus the filters made filtering work. I've tested by forcing a category (in this case "Recruitment"):

```js
data() { 
  return { 
    json: services, 
    error: null, 
    services: [], 
    loading: true, 
    buildDate: BUILD_DATE ? format(BUILD_DATE) : null, 
    commitHash: COMMIT_HASH, 
    show: ["maybe", "no"], 
    category: "Recruitment",
  }; 
}, 
```

And things were working perfectly given I could filter by paid/free/maybe AND then filter by a category, (in this case "Recruitment"). Exactly the behaviour I wanted. **Having said that, I'm now stuck with making it work for a category chosen from the UI.**

i.e. from choosing it from the list of categories in the UI does not filter anything:

<img width="1152" alt="Screenshot 2020-05-18 at 22 57 10" src="https://user-images.githubusercontent.com/41002226/82264632-31848c80-995d-11ea-8d32-e09bfe496acd.png">

--> link to the UI (but rendering may stop and you'd get a blank page once a a new github action runs / see my previous comment about Github Pages relative paths): [https://codeconomy.github.io/codeconomy.github.io-list/](https://codeconomy.github.io/codeconomy.github.io-list/)

I have been attempting a few things to tie the filter selection to what is returned by `showServices` but haven't been lucky.

I'd welcome any pointers :)

Thanks a lot again! Really appreciate the guidance.

**Comment by codeconomy** [comment]
> I have been attempting a few things to tie the filter selection to what is returned by `showServices` but haven't been lucky.
> 
> I'd welcome any pointers :)

Managed to get it to work. 🙌🏼 Will prepare add PR from my active account. disclaimer: very non-DRY code ...

Opening a new issue in a second to ask whether you have a suggestion to generate unique pages to each of the listings according to a template (i.e. not linking out to the different SaaS products (in your case), but to pages in saasforcovid.com that would offer more details about each saas product.)

Thanks a lot for all the help @adriaanvanrossum

**Comment by adriaandotcom** [comment]
You did a great job! DRY is not important. People using the product wouldn't know :)

One little thing: it seems like two categories are not working at the moment:

![image](https://user-images.githubusercontent.com/1079135/82577580-77269c80-9b8b-11ea-8e54-414767677919.png)

**Comment by codeconomy** [comment]
Thanks for testing @adriaanvanrossum ! :)

Just saw the typo leading to it. Will send a PR.
